OpenSim/Common/Reporter.cpp

```diff
@@ -0,0 +1,120 @@
+ *                                                                            *
+ * Unless required by applicable law or agreed to in writing, software        *
+ * distributed under the License is distributed on an "AS IS" BASIS,          *
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.   *
+ * See the License for the specific language governing permissions and        *
+ * limitations under the License.                                             *
+ * -------------------------------------------------------------------------- */
+
+#include "Reporter.h"
+#include <OpenSim/Common/TimeSeriesTable.h>
+
+using namespace SimTK;
+
+namespace OpenSim {
+
+class OutputPeriodicReporter : public SimTK::PeriodicEventReporter {
+public:
+    OutputPeriodicReporter(const AbstractReporter& owner,
+        const MultibodySystem& system,
+        const Real reportInterval) : PeriodicEventReporter(reportInterval), 
+        _owner(owner), _system(system) {
+    }
+
+    void handleEvent(const State& state) const override {
+        // This should be triggered every (interval) time units.
+        SimTK_ASSERT(state.getTime() == getNextEventTime(state, true),
+            "Reporter did not report at specified time interval.");
+
+        // delegate back to the OpenSim::Reporter to do the reporting
+        _owner.report(state);
+    }
+
+private:
+    const AbstractReporter& _owner;
+    const MultibodySystem& _system;
+};
+
+//=============================================================================
+// CONSTRUCTOR(S) AND DESTRUCTOR
+//=============================================================================
+//_____________________________________________________________________________
+// Default constructor.
+AbstractReporter::AbstractReporter()
+{
+    setNull();
+    constructProperties();
+}
+
+AbstractReporter::AbstractReporter(SimTK::Xml::Element& node) : Super(node)
+{
+    setNull();
+    constructProperties();
+}
+
+
+
+//=============================================================================
+// CONSTRUCTION METHODS
+//=============================================================================
+
+//_____________________________________________________________________________
+// Set the data members of this Reporter to their null values.
+void AbstractReporter::setNull()
+{
+    setAuthors("Ajay Seth");
+    setName("reporter");
+}
+
+//_____________________________________________________________________________
+// Define properties.
+void AbstractReporter::constructProperties()
+{
+    constructProperty_is_disabled(false);
+    constructProperty_report_time_interval(0.0);
+}
+
+// Create an underlying SimTK::Reporter to represent the OpenSim::Reporter in the 
+// computational system.  Create a SimTK::Reporter::Custom by default.
+void AbstractReporter::extendAddToSystem(SimTK::MultibodySystem& system) const
+{
+    Super::extendAddToSystem(system);
+    if (get_is_disabled())
+        return;
```

Comment:
I think this check should maybe also occur in `extendRealizeReport()`. As is, the reporter will still spew stuff out if not using a reporting interval. This can then be revisited once we are using a proper Event to handle the reporting at each successful step.